File: Week_03/Practicum/Find First and Last Position of Element in Sorted Array.cpp
// https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/description

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        // Find the first occurrence of the target using lower_bound
        // lower_bound returns the iterator to the first element >= target
        int low = lower_bound(nums.begin(), nums.end(), target) - nums.begin();

        // If the target is not found in the array, return {-1, -1}
        // Either we reached the end of the array, or the element at "low" is not the target
        if (low == nums.size() || nums[low] != target) {
            return {-1, -1};
        }

        // Find the last occurrence of the target using upper_bound
        // upper_bound returns the iterator to the first element > target
        // Subtracting 1 gives us the last index of the target
        int up = upper_bound(nums.begin(), nums.end(), target) - nums.begin() - 1;

        // Return the starting and ending index of the target in the array
        return {low, up};
    }
};

File: Week_03/Practicum/python/Find First and Last Position of Element in Sorted Array.py
import bisect

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if len(nums) == 0:
            return -1, -1

        idx1, idx2 = bisect.bisect_left(nums, target), bisect.bisect_right(nums, target) - 1
        if idx1 == len(nums) or nums[idx1] != target:
            return -1, -1

        return idx1, idx2